src/kaleidoscope/plugin/OneShot.cpp

```diff
@@ -22,266 +22,339 @@
-        !(mapped_key.getFlags() == (KEY_FLAGS | SYNTHETIC | SWITCH_TO_KEYMAP))) {
-      should_cancel_ = true;
+  } else {
+    if (state_[n].active && !state_[n].sticky) {
+      // If a OneShot key is held (on its initial press) long enough
+      // (i.e. > `hold_time_out` milliseconds), deactivate it. It
+      // still remains in effect until it is released.
+      if (Runtime.hasTimeExpired(start_time_, hold_time_out)) {
+        state_[n].active = false;
+      }
     }
   }
 
+  // Finally, we allow event processing to continue so that plugins
+  // later in the processing order can act on OneShot key events (in
+  // particular, ActiveModColor).
   return EventHandlerResult::OK;
 }
 
+
+// For any active OneShot modifier keys, keep those modifiers active
+// in the keyboard HID report.
 EventHandlerResult OneShot::beforeReportingState() {
-  for (uint8_t i = 0; i < ONESHOT_KEY_COUNT / 2; i++) {
-    if (state_[i].active) {
-      activateOneShot(i);
+  for (uint8_t i{0}; i < ONESHOT_KEY_COUNT; ++i) {
+    if (state_[i].active || state_[i].pressed) {
+      sustainOneShot(i);
     }
   }
-
   return EventHandlerResult::OK;
 }
 
+
 EventHandlerResult OneShot::afterEachCycle() {
-  bool oneshot_active = false;
-  for (uint8_t i = 0; i < ONESHOT_KEY_COUNT; i++) {
-    if (state_[i].active) {
-      oneshot_active = true;
-      break;
-    }
-  }
-  if (oneshot_active && hasTimedOut())
-    cancel();
-
-  bool is_cancelled = false;
-
-  for (uint8_t i = 0; i < ONESHOT_KEY_COUNT; i++) {
-    if (should_cancel_) {
-      if (state_[i].sticky) {
-        if (should_cancel_stickies_) {
-          is_cancelled = true;
-          state_[i].sticky = false;
-          cancelOneShot(i);
-          state_[i].pressed = false;
-        }
-      } else if (state_[i].active && !state_[i].pressed) {
-        is_cancelled = true;
-        cancelOneShot(i);
+  // If a normal, non-modifier key has been pressed, or if active,
+  // non-sticky OneShot keys have timed out, this is where they get
+  // released. Release is triggered when `release_countdown_` gets to
+  // 1, not 0, because most of the time it will be 0 (see below). It
+  // gets set to 2 on the press of a normal key when there are any
+  // active OneShot keys; that way, the OneShot keys will stay active
+  // long enough to apply to the newly-pressed key.
+  if ((release_countdown_ == 1) || hasTimedOut()) {
+    for (uint8_t i{0}; i < ONESHOT_KEY_COUNT; ++i) {
+      if (state_[i].active && !state_[i].sticky && !state_[i].pressed) {
+        releaseOneShot(i);
       }
     }
   }
-
-  if (is_cancelled) {
-    should_cancel_ = false;
-    should_cancel_stickies_ = false;
-  }
+  // Also, advance the counter for OneShot keys that have been
+  // cancelled by the press of a non-OneShot, non-modifier key. An
+  // unconditional bit shift should be more efficient than checking
+  // for zero to avoid underflow.
+  release_countdown_ >>= 1;
```

Comment:
This is perhaps too "clever" to be easily understood. I hope the comment I wrote is adequate.